### ISSUE TITLE
AtlasEngine rendering wide Nerd Font glyphs incorrectly

### ISSUE THREAD

**Issue by SivanagBalla** [microsoft/terminal#14022]
### Windows Terminal version

 1.16.2523.0

### Windows build number

10.0.19044

### Other Software

UbuntuMono Nerd font installed from https://github.com/ryanoasis/nerd-fonts/releases/download/v2.2.2/Ubuntu.zip
WSL, running Ubuntu 20.04.1 LTS. SSHing into a RHEL server
The linux server is running tmux with custom status bar with "double wide" icon characters

### Steps to reproduce

Use UbuntuMono Nerd Font and configure some double wide character in status bar or print them in terminal 

Issue is seen with new AtlasEngined enabled(which is default with new version). Disabling AtlasEngine will bring back old behaviour

### Expected Behavior

Double wide characters such as clock, calender which are available in NerdFond are to be rendered as below

![image](https://user-images.githubusercontent.com/19350704/190709997-35fe81c5-d140-4de1-b12a-ec1233cb8fc7.png)
![image](https://user-images.githubusercontent.com/19350704/190710041-609808cf-10d9-415c-8f35-3a46a0252df8.png)

### Actual Behavior

![image](https://user-images.githubusercontent.com/19350704/190710323-5d31dd7f-edd3-41f5-8028-57b81f607d6c.png)
![image](https://user-images.githubusercontent.com/19350704/190710363-7cd8eed9-ce04-472e-b560-f43020675edd.png)

**Comment by T4P4N** [comment]
same here

**Comment by lhecker** [comment]
I gave this issue a title that will hopefully make it easier to find in the future. 🙂

FYI I'm currently working on a version of AtlasEngine that solves this issue fundamentally. My prototype is already capable of handling all Nerd Font glyphs correctly. Unfortunately however it comes with its own set of downsides and is a large change, so the jury is still out on when and in what shape we'll fix this issue. Don't worry though: Until it's been fixed, the old text renderer will not go away.

**Comment by stephc-int13** [comment]
> I gave this issue a title that will hopefully make it easier to find in the future. 🙂
> 
> FYI I'm currently working on a version of AtlasEngine that solves this issue fundamentally, by going with a more traditional text rendering approach, allowing for freely overlapping glyphs. My prototype is already capable of handling all Nerd Font glyphs correctly. Unfortunately however it comes with its own set of downsides (slightly slower, more complex, doesn't support user-defined text shaders), so the jury is still out on when and in what shape we'll fix this issue. Don't worry though: Until it's been fixed, the old text renderer will not go away.

You could probably do that in two passes.
First pass with the usual shader rendering standard cells, second pass rendering the overlapping cells if necessary.
That is probably optimal and avoid unnecessary overhead.

**Comment by lhecker** [comment]
> You could probably do that in two passes.
> First pass with the usual shader rendering standard cells, second pass rendering the overlapping cells if necessary.
> That is probably optimal and avoid unnecessary overhead.

That would mean I'd have to keep both, the complexity of our current approach and add the complexity of the new approach for fallback. That would probably make maintainability more difficult than a larger one time change. An idea I had to keep some of the current benefits was to just composite all text into a viewport-sized offscreen texture and blend it onto the swap chain in a single final shading pass, similar to how we do it now.
BTW to give some context on the "slightly slower" / overhead I mentioned: The new approach I've prototyped runs within my test setup at ~200us per frame, whereas the old one needs ~100us per frame. Sure, that's twice as slow, but I'm not even using instancing yet to render glyph quads (or Nvidia's `QuadFillMode`), nor are we anywhere close to the 500-1000us that the old renderer needs. 😅
I still got quite a number of kinks to iron out, but once that's done I'll discuss with everyone on the team how to best integrate such changes.

![image](https://user-images.githubusercontent.com/2256941/192152695-b41f609b-557d-4da6-ad22-2163496289e5.png)

**Comment by stephc-int13** [comment]
I don't think you need to fall back on per glyph rendering to solve the special wide glyphs needed for proper nerdfont icons rendering. 

I understand that the speed difference is not significant, IMHO your solution is more complex than doing two passes with shader on a quad.

First pass as already done. (Ignoring wider glyphs)
Second pass only on "special" glyphs with wider cells.

It might even be possible to do everything in a single pass shader, but that should be tested.

**Comment by lhecker** [comment]
When we render we need to conceptually draw in "layers" from bottom to top (including things that are upcoming):
* 1st pass:
	* background color
	* selection (variant 1)
	* cursor (variant 1 and 2)
	* glyphs
	* text decorations (overline, underline, strikethrough, etc.)
	* sixels, etc.
	* selection (variant 2)

This can be done in a single compute shader pass. But in order to draw wide glyphs in between we don't have to draw in 2 passes, but in 3, since things like sixels or selections are "on top" of the lower contents. From what I can see it basically turns into:
* 1st pass:
	* background color
	* selection (variant 1)
	* cursor (variant 1 and 2)
	* glyphs
	* † text decorations (overline, underline, strikethrough, etc.)
	* † sixels, etc.
	* † selection (variant 2)
* 2nd pass (optional):
	* wide glyphs
	* sixels, etc.
* 3rd pass ‡:
	* ‡ text decorations (overline, underline, strikethrough, etc.)
	* ‡ sixels, etc.
	* ‡ selection (variant 2)

† optional if no 2nd pass
‡ required if 2nd pass

This is my new approach:
* 1st pass:
	* background color
	* selection (variant 1)
* 2nd pass:
	* glyphs
	* text decorations (overline, underline, strikethrough, etc.)
	* sixels, etc.
	* selection (variant 2)

Additionally it actually reduces the amount of code required for text rendering by ~12% (it's just that the remaining code requires more intimate knowledge about DirectWrite and Direct3D). As such right now I personally don't agree that keeping both approaches simultaneously offers a strong benefit.
I'm open to any ideas how to not draw individual quads and I'd be happy to be convinced otherwise, but even if you're right that it's simpler and better, given _my_ knowledge I just wouldn't know how that would work.

**Comment by stephc-int13** [comment]
The more I think about it the more I think that the whole rendering can be entirely done in one single shader pass.

But that does require a bit of preprocessing and adding metadata for each cell to handle all the cases.
That would mostly be a few special case flags to tell the shader to look for more information in adjacent cells (top, left or bottom) 

But of course, given that performance gains are not critical (the optimal solution would be a bit more power efficient) you should simply follow your guts and go for the solution that is the simpler to implement _for you_.

**Comment by lhecker** [comment]
Thanks for your input! I'll definitely try my hardest to come up with something decent. Nothing is set in stone so far. 🙂

**Comment by jessey-git** [comment]
I seem to have a similar problem, except that version 2523 is ok, but the most recent 2642 is broken. I'm using MesloLGM NF at size 9.0 (tried 8.5 and 9.5 as well and those were much worse).

2523 on top, 2642 on bottom:
![fontbrokenness](https://user-images.githubusercontent.com/7989986/192405105-9fd1c813-71e5-4e54-a91a-58dc642236c4.png)

Is this the same or a different issue?

**Comment by stephc-int13** [comment]
You can disable the new atlas renderer in the settings, if that solves your issue then it is related.

**Comment by jessey-git** [comment]
Yeah, disabling Atlas fixes the problem. Welp, it worked for almost 14 days with the prior version. It had a good run :)  Will try Atlas again once this issue here is all fixed.

**Comment by lhecker** [comment]
@jessey-git I'm sorry I broke your terminal! 😔

I hope you don't mind me saying this, but your font is... weird. 😄
It looks beautiful of course, but it handles its block characters in a very unusual manner. That character (U+E0B0) works in all other Nerd Fonts I got installed (for instance Fira Code), just not with this Meslo variant, which is already a bit weird. Additionally my understanding is that block characters are expected to have approximately the same size as the font's advance width and height. While most Nerd Fonts variants are slightly incorrect in that regard, Meslo in particular is _really_ bad about this. This particular glyph for instance is almost 40% taller than the advance height it specifies which causes it to look like that. The only difference between build 2523 and 2642 is how those glyphs are vertically centered. In both screenshots you're missing more than half of that glyph - it's actually supposed to be a perfect triangle like this:
![image](https://user-images.githubusercontent.com/2256941/192419277-64440875-137a-4aa0-9e0a-fda2df68a1d5.png)

Basically, IMO Nerd Fonts should stop making block glyphs taller than the font's advance height.
In either case I'm going to simply revert the change that causes your issues: #14085
